### ISSUE TITLE
Allow manage database connections

### ISSUE THREAD

**Issue by strike** [coder/coder#11897]
**TLDR: Allow configuring Database connection parameters for improve performance**

Recently, we encountered slow connection issues and identified a problem of extremely slow responses, despite the Database Management System (DBMS) and backend having low load. Using OpenTelemetry (OTEL), we observed the following scenario:

```plaintext
Query duration: 1 minute
Database responses: Fast (in out non-standard DBMS configuration in another DC, average delay around 100ms)
CPU load: Insignificant (less than 2 cores)
```

![image](https://github.com/coder/coder/assets/223260/59ea8464-69b9-4c42-ad1d-ba290f509f08)

A significant portion of queries experienced timeouts. Upon investigation, we found a hard-coded limitation severely restricting the number of connections to the DBMS:

- `SetMaxOpenConns = 10`
- `SetMaxIdleConns = 3`

We recompiled the application with increased limits (`SetMaxOpenConns = 200`, `SetMaxIdleConns = 60`) and conducted a simple stress test (500 simultaneous requests):

```bash
ab -n 500 -c 500 -H "Coder-Session-Token: $CODER_SESSION_TOKEN" coder/api/v2/workspaces
```

Before the patch:

```plaintext
$ ab -n 500 -c 500 -s 30 -H "Coder-Session-Token: $CODER_SESSION_TOKEN" https://coder/api/v2/workspaces
This is ApacheBench 
...
 100%  119864 (longest request)
```

The longest request took around 120 seconds.

After the patch:

```plaintext
$ ab -n 500 -c 500 -H "Coder-Session-Token: $CODER_SESSION_TOKEN" https://coder/api/v2/workspaces
This is ApacheBench 
...
 100%  11604 (longest request)
```

The longest request now takes around 12 seconds. However, performance is now limited by CPU, with a set ceiling of 20 CPUs in the tests.

**Comment by kylecarbs** [comment]
@coadler

**Comment by coadler** [comment]
Yeah, this is definitely something we should allow configuration for. In general though, Coder expects <15ms database latency. I'd expect even with these higher connection limits for the experience to be poor all around either way. Is there a reason your Coder deployment needs to be in a different region?

**Comment by strike** [comment]
Yes, we are aware that it's preferable to have the backend as close as possible to the database. Unfortunately, right now we do not have the capability to do so, but I think we will implement such a scheme in the future.
By the way, except for the issue with `SetMaxOpenConns`, Coder is working quite well.
